File: gitlfs-client/src/test/java/ru/bozaro/gitlfs/client/HttpRecorder.java
package ru.bozaro.gitlfs.client;

import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpUriRequest;

import javax.annotation.Nonnull;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

/**
 * Client with recording all request.
 *
 * @author Artem V. Navrotskiy
 */
public class HttpRecorder implements HttpExecutor {
  @Nonnull
  private final HttpExecutor executor;
  @Nonnull
  private final List<HttpRecord> records = new ArrayList<>();

  public HttpRecorder(@Nonnull HttpExecutor executor) {
    this.executor = executor;
  }

  @Nonnull
  @Override
  public CloseableHttpResponse executeMethod(@Nonnull HttpUriRequest request) throws IOException {
    final CloseableHttpResponse response = executor.executeMethod(request);
    records.add(new HttpRecord(request, response));
    return response;
  }

  @Nonnull
  public List<HttpRecord> getRecords() {
    return records;
  }

  @Override
  public void close() throws IOException {
    executor.close();
  }
}

File: gitlfs-client/src/test/kotlin/ru/bozaro/gitlfs/client/HttpRecorder.kt
package ru.bozaro.gitlfs.client

import org.apache.http.client.methods.CloseableHttpResponse
import org.apache.http.client.methods.HttpUriRequest
import java.io.IOException

/**
 * Client with recording all request.
 *
 * @author Artem V. Navrotskiy
 */
class HttpRecorder(private val executor: HttpExecutor) : HttpExecutor {
    val records = ArrayList<HttpRecord>()

    @Throws(IOException::class)
    override fun executeMethod(request: HttpUriRequest): CloseableHttpResponse {
        val response = executor.executeMethod(request)
        records.add(HttpRecord(request, response))
        return response
    }

    @Throws(IOException::class)
    override fun close() {
        executor.close()
    }
}
